### ISSUE TITLE
Genomics API fails to decode ImportVariantsResponse

### ISSUE THREAD

**Issue by alexherns** [google/google-api-go-client#157]
Calls to the v1 Genomics API using this library fail to correctly decode the Response fields of an Operation, such as when importing Variants. `Operation` objects returned from `(*VariantsImportCall) Do` have a `Response` field that cannot be cast as `ImportVariantsResponse`, but are instead of type `map[string]interface{}`. Marshaling an `Operation` to JSON and unmarshalling back into the following structs yields the correct types, but I don't know how to fix the internal decoder to remedy this behavior:

```go
type importOperationResponse struct {
	Response importVariantsResponse `json:"response"`
}

type importVariantsResponse struct {
	CallSetIds []string `json:"callSetIds"`
}
```

**Comment by mcgreevy** [comment]
Unfortunately there's no good way of making the generator do the right thing here, because the necessary information is not present in the _discovery doc_ which defines the API.  I'm drafting a list of possible ways we could attack this, which I will post tomorrow, but each has its own shortcomings.

**Comment by mcgreevy** [comment]
@bradfitz 

The issue here is that the Response field is of type OperationResponse, which is defined as:

type OperationResponse interface{}

The comment on the Response field specifies which concrete types the field may contain, but the JSON discovery doc doesn't actually provide any info for the generator to know the possible types. I don't see a way to generate code which automatically unmarshals into the correct concrete type. 
It looks like the returned JSON data for these "any" fields almost always[1] has an "@type" field which contains a type URI specifying which type has actually been returned (i.e. whether it's a ImportVariantsResponse or a ImportReadGroupSetsResponse in this case), but it's not clear how to associate the @type with the Go type (or the schema name in the discovery doc).

I think our best option here (should we do anything) is to define a type --
type Any interface{}
-- to replace all of the "interface{}" types like "OperationResponse".  This could contain the same value as it does now, for backwards compatability.  We then add an "Unmarshal" (name TBD) method on Any, which takes an interface provided by the user and does the round-trip via JSON that alexherns refers to.  We might be able to also provide another method on Any which returns the @type value, to help the user decide which concrete type to pass to Unmarshal.

I'm not sure how much of a win this is. It would do some of the dirty work for users of the APIs, but it would be still be difficult to document how to use this.  For the Genomics API case we wouldn't, for example, be able to automatically generate documentation to say that
  operation.Response.Unmarhsal(v interface{})
should be given either a *ImportVariantsResponse or a *ImportReadGroupSetsResponse.  The doc string happens to mention that, but in other similar cases the user would still be in the dark.

[1] There are a few APIs which use raw JSON (not associated with any schema type) for a few fields.  The main one is the generated logging API, with its LogEntry.StructPayload field.  There are also a few of these fields in the dataflow api.

**Comment by okdave** [comment]
@mcgreevy to clarify, under the hood the `Any` is going to hold the same `map[string]interface{}` that is stuffed into the `OperationsResponse` today? Or do you think we should do something more aggressive (like hold the raw `[]byte` of the JSON and make users call `.Unmarshal(myMap)` if they want the same value as today?

**Comment by mcgreevy** [comment]
> @mcgreevy to clarify, under the hood the Any is going to hold the same map[string]interface{} that is stuffed into the OperationsResponse today

Yes, I think it should, to minimize disruption to existing user code.  I was concerned about the round-trip JSON marshaling causing large ints to lose precision (as they would be unmarshalled into float64s), but it looks like large ints are generally stored as strings, using the string json struct tag, so I don't think it's actually an issue.

To be clear, however, I'm not convinced that this proposal is a huge advance over the status quo. It's just the least-bad option that I can see.  I'm still unsure how we would effectively document the "Any.Unmarshal" method.  I'm interested to hear @bradfitz's take.
